src/databricks/labs/ucx/hive_metastore/table_migrate.py

```diff
@@ -18,61 +19,52 @@ def __init__(
+        mapping_rules = self._get_mapping_rules_dict()
         tasks = []
         for table in self._tc.snapshot():
-            target_catalog = self._default_catalog
-            if self._database_to_catalog_mapping:
-                target_catalog = self._database_to_catalog_mapping[table.database]
-            tasks.append(partial(self._migrate_table, target_catalog, table))
-        _, errors = Threads.gather("migrate tables", tasks)
-        if len(errors) > 0:
-            # TODO: https://github.com/databrickslabs/ucx/issues/406
-            # TODO: pick first X issues in the summary
-            msg = f"Detected {len(errors)} errors: {'. '.join(str(e) for e in errors)}"
-            raise ValueError(msg)
-
-    def _migrate_table(self, target_catalog: str, table: Table):
-        sql = table.uc_create_sql(target_catalog)
-        logger.debug(f"Migrating table {table.key} to using SQL query: {sql}")
-        target = f"{target_catalog}.{table.database}.{table.name}".lower()
-
-        if self._table_already_upgraded(target):
-            logger.info(f"Table {table.key} already upgraded to {self._seen_tables[target]}")
-        elif table.object_type == "MANAGED":
+            rule = mapping_rules.get(table.key)
+            if not rule:
+                logger.info(f"Skipping table {table.key} table doesn't exist in the mapping table.")
+                continue
+            tasks.append(partial(self._migrate_table, table, rule.as_uc_table_key))
+        Threads.strict("migrate tables", tasks)
+
+    def _migrate_table(self, src_table: Table, target_table_key):
+        sql = src_table.uc_create_sql(target_table_key)
+        logger.debug(f"Migrating table {src_table.key} to using SQL query: {sql}")
+
+        if self._table_already_upgraded(target_table_key):
+            logger.info(f"Table {src_table.key} already upgraded to {target_table_key}")
+            return True
+        if src_table.object_type == "MANAGED":
             self._backend.execute(sql)
-            self._backend.execute(table.sql_alter_to(target_catalog))
-            self._backend.execute(table.sql_alter_from(target_catalog))
-            self._seen_tables[target] = table.key
-        elif table.object_type == "EXTERNAL":
-            result = next(self._backend.fetch(sql))
-            if result.status_code != "SUCCESS":
-                raise ValueError(result.description)
-            self._backend.execute(table.sql_alter_to(target_catalog))
-            self._backend.execute(table.sql_alter_from(target_catalog))
-            self._seen_tables[target] = table.key
-        else:
-            msg = f"Table {table.key} is a {table.object_type} and is not supported for migration yet"
-            raise ValueError(msg)
-        return True
+            self._backend.execute(src_table.sql_alter_to(target_table_key))
+            self._backend.execute(src_table.sql_alter_from(target_table_key))
+            self._seen_tables[target_table_key] = src_table.key
```

Comment:
BUG: `self._seen_tables` is modified from multiple threads and it would cause broken behavior. Compute it once before multiple threads are started.

Reply:
KK

Reply:
Fixed